Support Animations in Model64

gamemasterplc opened the pull request:
Added support for animations in model64. Up to 3 active animations can be played at once if they do affect different bones.

gamemasterplc committed:
Implement animation data output

gamemasterplc committed:
Add data load functions to model64

gamemasterplc committed:
Fix reference count with model64_create

gamemasterplc committed:
Get basic animations working

Animations are still buggy especially when streaming

gamemasterplc committed:
Fix streaming animation data from ROM

gamemasterplc committed:
Fix animation playback

gamemasterplc committed:
Add better safety checking for slerp in mkmodel

gamemasterplc committed:
Use true slerp for quaternion interpolation

Fixes 1-frame glitches during animations

gamemasterplc committed:
Add extra frame to animations

Makes animations have maximum time as part of them always

gamemasterplc committed:
Make animation speed 0 not update node transforms

gamemasterplc committed:
Add model64_anim_get_length

gamemasterplc committed:
Change runtime quaternion lerp to nlerp

gamemasterplc committed:
Allow for 3 active animations at once

gamemasterplc committed:
Merge branch 'DragonMinded:unstable' into unstable

gamemasterplc committed:
Merge branch 'DragonMinded:unstable' into unstable

gamemasterplc committed:
Add documentation to undocumented model64 data types

Fixes doxygen build

gamemasterplc committed:
Fix short animation wrapping

gamemasterplc committed:
Allow sub-frame duration for animation

gamemasterplc committed:
Invalidate animation pose when starting new animation

Fixes starting paused animation.

gamemasterplc committed:
Change animation API to take in explicit slot number

gamemasterplc committed:
Make animation slot an enum

snacchus commented on src/model64.c:
```diff
@@ -464,3 +524,207 @@ void model64_draw(model64_t *model)
+void model64_anim_play(model64_t *model, const char *anim, model64_anim_slot_t slot, bool paused, float start_time)
+{
+    int32_t anim_index = search_anim_index(model, anim);
+    assertf(anim_index != -1, "Invalid animation name");
```
I'd suggest a more descriptive error message, e.g. "Animation '%s' was not found"

snacchus commented on src/model64.c:
```diff
@@ -464,3 +524,207 @@ void model64_draw(model64_t *model)
+    int32_t anim_index = search_anim_index(model, anim);
+    assertf(anim_index != -1, "Invalid animation name");
+    if(model->data->stream_buf_size != 0 && !model->active_anims[slot].stream_buf[0]) {
+        model->active_anims[slot].stream_buf[0] = malloc_uncached(model->data->stream_buf_size);
+        model->active_anims[slot].stream_buf[1] = malloc_uncached(model->data->stream_buf_size);
```
These buffers are read on CPU, right? In that case I would allocate them with cached memory (and invalidate them accordingly), at least for now. If we offload the processing to the RSP at some point we can switch to uncached again when the time comes.

gamemasterplc replied:
The buffers are definitely read on the CPU.

snacchus commented on src/model64.c:
```diff
@@ -464,3 +524,207 @@ void model64_draw(model64_t *model)
+float model64_anim_get_length(model64_t *model, const char *anim)
+{
+    int32_t anim_index = search_anim_index(model, anim);
+    assertf(anim_index != -1, "Invalid animation name");
```
See my other comment in `model64_anim_play`

snacchus commented on src/model64.c:
```diff
@@ -464,3 +524,207 @@ void model64_draw(model64_t *model)
+        dma_read(anim_slot_ptr->stream_buf[0], rom_addr+(curr_data_idx*frame_size), frame_size);
+        buf_info->curr_buf = anim_slot_ptr->stream_buf[0];
+        dma_read(anim_slot_ptr->stream_buf[1], rom_addr+(next_data_idx*frame_size), frame_size);
+        buf_info->next_buf = anim_slot_ptr->stream_buf[1];
```
Could we reuse previously loaded data here? For example if `stream_buf[1]` already contains the current frame (because it was the "next frame" previously), we don't need to DMA it again. And the other way around of course in case time is moving backwards.

gamemasterplc replied:
Unfortunately, reusing the existing buffers is pretty hard without copying via DMA or the CPU.

snacchus replied:
Why is it hard? You already have a level of indirection that you could use for this. You could swap the values of `buf_info->curr_buf` and `buf_info->next_buf` if required. Unless I'm missing something?

gamemasterplc replied:
Unfortunately, I have trouble getting interpolation between read keyframes to work correctly when reusing streaming buffers which results in visible shaking or jerking to keyframes especially at slow speeds. Btw, I'm also trying to skip both the current and next fetch when the frame number hasn't changed.

snacchus commented on src/model64.c:
```diff
@@ -464,3 +524,207 @@ void model64_draw(model64_t *model)
+            model->active_anims[i].new_pose = false;
+            continue;
+        }
+        model->active_anims[i].time += model->active_anims[i].speed*dt;
```
Do we need to worry about accumulating error here? From what I understand, Users can set the current time freely via `model64_anim_set_time` anyway, so why does this function take a delta? If it took the absolute time instead we could let users handle calculating it, to whatever precision they require.

gamemasterplc replied:
model64_anim_set_time is for seeking within the animation. The function takes in a delta for advancing the time of all active animations.

snacchus commented on src/model64.c:
```diff
@@ -464,3 +524,207 @@ void model64_draw(model64_t *model)
+        if(model->active_anims[i].loop) {
+            model64_anim_t *curr_anim = &model->data->anims[model->active_anims[i].index];
+            model->active_anims[i].time = fm_fmodf(model->active_anims[i].time, curr_anim->duration);
+            if(model->active_anims[i].time < 0) {
```
In which case could `time` be negative here? If speed is negative?

gamemasterplc replied:
Time can be negative if speed is negative because fmodf returns a negative value if time is negative. Negative speed can also cause time to go below 0.

snacchus commented on src/model64_internal.h:
```diff
@@ -88,19 +109,43 @@ typedef struct model64_data_s {
+    void *curr_buf;     ///< Data buffer for current frame
+    void *next_buf;     ///< Data buffer for next frame
+    float time;         ///< Interpolation factor between current and next frame
+} anim_buf_info_t;
```
This struct could be moved to model64.c since it is only used there as an implementation detail

snacchus commented on include/model64.h:
```diff
@@ -19,9 +30,13 @@ typedef struct primitive_s primitive_t;
 struct model64_node_s;
 typedef struct model64_node_s model64_node_t;
 
+model64_data_t *model64_load_data(const char *fn);
```
There was already a discussion about this on Discord, but here is my take: I can see the point in separating the model data ("blueprint"?) from model "instances". If we were to go that way, I'd double down on the separation and remove `model64_clone`, `model64_load` and `model64_load_buf`. That way it's very clear how to load a model (`model64_load_data`) and how to create instances (`model64_create`). Otherwise there would seem to be many ways of achieving the same thing (creating a model instance), which would lead to confusion IMO. Also I'd work on the naming a bit more.

However, I must say that I also agree with @rasky in that this is just increasing the API surface with functions that kind of do the same things as the already existing ones. The only actual difference is that with the extra functions, you can load model data without actually creating an instance, which would save some memory on the order of a couple of kilobytes, depending on how many nodes the model contains. We can discuss whether that is worth it. OTOH, with the changes I suggested above, we'd go from 4 functions to 5 in total, so maybe it would actually be fine.

gamemasterplc replied:
So, should I just remove the existing clone and plain load/load_buf functions. What should I rename the other functions to?

gamemasterplc replied:
Privatized these functions for now.

snacchus commented on src/model64_internal.h:
```diff
@@ -88,19 +109,43 @@ typedef struct model64_data_s {
-    node_transform_state_t *transforms;     ///< List of transforms for each bone in a model instance
+    model64_data_t *data;                           ///< Pointer to the model data this instance refers to
+    node_transform_state_t *transforms;             ///< List of transforms for each bone in a model instance
+    anim_state_t active_anims[MAX_ACTIVE_ANIMS];    ///< List of active animations
```
I think the animation state should be heap allocated too, like the transforms. Otherwise we are always allocating space for it even if a model doesn't contain animations.

gamemasterplc replied:
Would each animation state be heap allocated or just the whole list of active animations. The stream buffers are still only allocated when playing the first animation on a slot even when unchanged.

snacchus commented on tools/mkmodel/mkmodel.c:
```diff
@@ -72,8 +88,10 @@ void print_args( char * name )
-    fprintf(stderr, "   -o/--output <dir>         Specify output directory (default: .)\n");
-    fprintf(stderr, "   -v/--verbose              Verbose output\n");
+    fprintf(stderr, "   -o/--output <dir>       Specify output directory (default: .)\n");
+    fprintf(stderr, "   --anim-no-stream        Output animation data for not streaming.\n");
```
I'd suggest "Disable animation streaming" as the description

gamemasterplc committed:
Make some simple changes

Improve assert in model64_anim_play. Allocate streaming buffers in cached memory. Move anim_buf_info_t to model64.c. Improve --anim-no-stream description.

gamemasterplc committed:
Dynamically Allocate Animation States

gamemasterplc committed:
Privatize data load functions

rasky commented on src/model64.c:
```diff
@@ -545,6 +541,25 @@ static int32_t search_anim_index(model64_t *model, const char *name)
+{
+    if(model->active_anims[slot]) {
+        return;
+    }
```
In general, this style of coding is slower on N64 because you pay the overhead of a function call (which includes switching to a different instruction cache line) just to check whether there's something to do. It would be better to move this check into all callers (and even mark it as unlikely) to make sure it impacts near zero during runtime

gamemasterplc committed:
Reduce unnecessary DMAs

Only DMAs if the frame has changed. Only does 1 DMA if the previous frame and next frame are 1 apart.

gamemasterplc committed:
Move model->active_anims[slot] check to callers

gamemasterplc committed:
Rename dt in model64_update

snacchus reviewed (approved):
Looks good to me
